File: designPatterns/Behavioral_patterns/Template/typescript/ConcreteClass2.ts
import AbstractClass from "./AbstractClass";

/**
 * Usually, concrete classes override only a fraction of base class' operations.
 */
export default class ConcreteClass2 extends AbstractClass {
  protected requiredOperations1(): void {
    console.log("ConcreteClass2 says: Implemented Operation1");
  }

  protected requiredOperation2(): void {
    console.log("ConcreteClass2 says: Implemented Operation2");
  }

  protected hook1(): void {
    console.log("ConcreteClass2 says: Overridden Hook1");
  }
}

File: designPatterns/Behavioral_patterns/Template/python/ConcreteClass2.py
from AbstractClass import AbstractClass


class ConcreteClass2(AbstractClass):
    """
    Usually, concrete classes override only a fraction of base class'
    operations.
    """

    def required_operations1(self):
        print("ConcreteClass2 says: Implemented Operation1")

    def required_operations2(self):
        print("ConcreteClass2 says: Implemented Operation2")

    def hook1(self):
        print("ConcreteClass2 says: Overridden Hook1")
